### ISSUE TITLE
Possible wrong handling of RFCTYPE_BCD in recently introduced fill type checks

### ISSUE THREAD

**Issue by soulchild** [SAP/node-rfc#52]
Since updating to the latest version of node-rfc, our BAPI calls are being rejected with the following error:

> Number expected when filling field AMOUNT of type 2

We dug into the source code to find out that this error is raised by the recently introduced [fill type check commit](https://github.com/SAP/node-rfc/commit/36db0c445bbd035685b47a854d53a2b618ad11e4). `RFCTYPE_BCD` is [handled](https://github.com/SAP/node-rfc/blob/master/src/rfcio.cc#L192) exactly like `RFCTYPE_FLOAT`, but as far as we know, `RFCTYPE_BCD` is supposed to be a Decimal, not a Float (we're using it to store money values). Thus, we believe `RFCTYPE_BCD` should expect a String, not a Number. 

In older versions, we were able to successfully pass the following attribute to the BAPI call:

```json
{
  "AMOUNT": "13.37"
}
``` 

which IMHO is the correct way to represent a currency value in JSON. With the latest version of the SDK this would need to be changed to:

```json
{
  "AMOUNT": 13.37
}
``` 

to pass the fill type checks. But representing money values with a float type is generally a very bad idea. I looked into other SAP RFC implementations (especially PyRFC), where RFCTYPE_BCD is correctly associated with Python's Decimal type. Due to the fact that Javascript does not have a built-in Decimal type[1], a string value is the only way to safely pass in money values. 

What do you think? 

[1] Of course, there are modules like [decimal.js](https://github.com/MikeMcl/decimal.js/)

**Comment by soulchild** [comment]
Besides the new fill type checks, something else seems to have changed between our working (admittedly very old) version of the library (I can only guess that it must be something like v0.1.4) and more recent ones: 

Going back as far as v0.1.6, `RFCTYPE_BCD` values suddenly need to be passed in as float, whereas in the older version, I was able to pass in strings (which, again, make more sense to me for currency types, if there is no explicit Decimal type). Otherwise they are being rejected by SAP. 

Can you elaborate on what has changed between these versions (v0.1.4 and earlier and v0.1.6) that might be causing this behaviour? Maybe some sort of implicit conversion which has been removed? 

Thanks!

**Comment by bsrdjan** [comment]
Hi @soulchild, thank you very much for catching this, the older versions were actually more correct here. 

Both string and number types can and shall be allowed for ABAP BCD and FLOAT inputs and, as you mentioned, from the rounding errors perspective, strings are clearly superior for BCDs.

According to [SAP NetWeaver RFC SDK Doxygen Documentation](https://websmp107.sap-ag.de/~sapidb/012002523100011227432017E.zip) linked here at the bottom http://service.sap.com/rfc-library, the RfcSetString/RfcGetString methods can also set/get BCD and FLOAT fields. 

Using `RfcSetString` instead of `RfcSetFloat` would make possible accepting JavaScript numbers, strings and even Decimal objects for example, as inputs. Here `PyRFC` example (need to test with PyRFC first):

```python

 elif typ == RFCTYPE_BCD or typ == RFCTYPE_FLOAT:
            # cast to string prevents rounding errors, for floats and decimals
            cValue = fillString(str(value))
            rc = RfcSetString(container, cName, cValue, strlenU(cValue), &errorInfo)
            free(cValue)
```
In Python, FLOATs and BCDs outputs can be converted to Python floats and Decimals, thus [no change here](https://github.com/SAP/PyRFC/blob/master/src/pyrfc/_pyrfc.pyx#L2071).

Following the same principle with `node-rfc`, numbers, strings or number objects (Decimal...) shall be allowed as ABAP FLOAT or BCD inputs, casted to string and set ("filed") using `RfcSetString`.

On output ("wrap") side, both FLOATs and BCDs can be read into strings, using `RfcGetString` and  FLOATs can be safely transformed into ECMAScript numbers. With BCD, output as string seem to be the only viable alternative and I would fix that way.

From the application perspective, practically any decimal lib could be used and even decimal objects could be passed to `node-rfc` but only BCD strings would be returned from ABAP.

Would fix in this direction work for you ?

**Comment by soulchild** [comment]
Glad I could be of help. That sounds like a good solution to me and would probably solve my problem. Thanks for taking the time!

**Comment by bsrdjan** [comment]
Please feel free to test with just published v0.1.16

**Comment by soulchild** [comment]
Thanks for fixing the issue in such short time! Initial tests look very promising, everything seems to be working as expected when passing in strings. If anything unexpected comes up, I'll reopen the issue.

**Comment by mploigtweber** [comment]
I think that can not be the solution.
What happens to BCDs, which are now all output as a string. How should you be able to use it in Javascript?

Here is one example from BAPI_PRODORD_GET_DETAIL Table OPERATION before and after:

![image](https://user-images.githubusercontent.com/2803448/47569219-5f495100-d933-11e8-8106-3351f8046cab.png)

and there are quite a few BAPIs that the BCDs use not just for currencies. With all the fields you can not calculate mathematically anymore ... they are just strings. Or should one parse this now every time?

Please help me?

**Comment by soulchild** [comment]
Strings are IMHO the safest representation, because JS does not have a dedicated Decimal type.

If the value represents a currency, I'd recommend stuffing the strings into [decimal.js](https://github.com/MikeMcl/decimal.js/) and work with the resulting objects exclusively to not introduce any rounding errors. 

For other values, like your `quantity` example, you could just use `Number.parseFloat()`. I understand that this requires some extra effort, but, again, I think it's the best possible way. The accidental change to floats also broke backwards-compatibility without announcing them.

**Comment by soulchild** [comment]
For instance:

```js
console.log(1.04-0.43)
```

yields:

```
0.6100000000000001
```

It's just not safe to calculate currencies with floats. There are lots of [posts](https://stackoverflow.com/questions/3730019/why-not-use-double-or-float-to-represent-currency/) on the web advising against it, and if it worked for you in the past, you just got lucky. 😉

Maybe there's a way to grab the SAP field definition and come up with some automated way to convert BCD values to floats?

**Comment by mploigtweber** [comment]
Nice theoretical example ... unfortunately none from practice. A simple parseFloat of strings would throw out the same "error". It is a pity that you have such an effort now because of old well-known Javascript properties. If the mistake had disturbed me at one point, I would have written

parseFloat ((1.04 - 0.43) .toPrecision (3))

>> 0.61 ... voila

Here are some examples. Also with decimal.js:

![image](https://user-images.githubusercontent.com/2803448/47601323-f1526780-d9ce-11e8-9da4-6a457015c1dd.png)

"Maybe there's a way to grab the SAP field definition and come up with some automated way to convert BCD values to floats?"

Can it be more complicated, please? 

Better way:

Maybe there's a way to grab the SAP field definition and come up with some automated way to convert BCD values to **strings**?

**Comment by soulchild** [comment]
I'm far from being an expert in data representation formats, but let's approach this from a different angle: IMHO a library like `node-rfc` is supposed to be more correct than your own applications because it's potentially used by quite a lot of people, who will be very confused if it causes some hard-to-find/debug rounding errors. Your examples might work for you, and I admit that probably most calculations with money floats will never cause any trouble in the real world. But still, I believe a library has to be more precise, even if this comes at the cost of being more unwieldy.

If you look at what a [BCD](https://en.wikipedia.org/wiki/Binary-coded_decimal) represents and how it represents it, we can see that it uses a fixed number of bits to represent **all** the numbers of a value. So there's no potential for rounding errors or imprecise representations of a given number.

[Floating point](https://en.wikipedia.org/wiki/Floating-point_arithmetic), on the other hand, uses an **approximation** of a value with a potentially infinite number of decimal digits.

So IMHO, the closest built-in data type to represent BCD values in Javascript is in fact a string. That does not necessarily mean that we should immediately stop thinking about whether it would make sense to have `node-rfc` return BCD values as float **optionally**.

Don't get me wrong: I hate nothing more than academic discussions about practical problems, but in this case, I firmly believe that the fix (which reverted a **breaking** change) was the right thing to do.

What do you think, @bsrdjan?

**Comment by soulchild** [comment]
I can totally understand that you're upset about this change because that's exactly what happened to us a while back when we upgraded from a very early version of `node-rfc` to a more recent one: Indeed the newer (minor) release suddenly returned floats instead of strings.

I hope, @bsrdjan joins the discussion so we can work something out which works for both of us (and other prospective users). :wink:

**Comment by bsrdjan** [comment]
Reading a BCD value, coming from **MD_STOCK_REQUIREMENTS_LIST_API** BAPI for example, starts with [reading it as a string](https://github.com/SAP/node-rfc/blob/napi/src/rfcio.cc#L589), from SAP NW RFC data container. That is the only rounding-safe way to get this value from ABAP:

```c++
    case RFCTYPE_BCD:
    {
        // An upper bound for the length of the _string representation_
        // of the BCD is given by (2*cLen)-1 (each digit is encoded in 4bit,
        // the first 4 bit are reserved for the sign)
        // Furthermore, a sign char, a decimal separator char may be present
        // => (2*cLen)+1
        unsigned int resultLen;
        unsigned int strLen = 2 * cLen + 1;
        SAP_UC *sapuc = mallocU(strLen + 1);
        rc = RfcGetString(functionHandle, cName, sapuc, strLen + 1, &resultLen, &errorInfo);
        if (rc != RFC_OK)
        {
            free(sapuc);
            break;
        }
        resultValue = wrapString(sapuc, resultLen);
        free(sapuc);
        break;
    }
```
At this point, at C++ and SAP NW RFC SDK Library level, only very restricted subset of node JS methods is provided by [node addon api](https://nodejs.github.io/node-addon-api/namespace_napi.html), we could use to parse this string. The length of input JavaScript string for example is at this level not any more available, type checks are restricted and extending node addon api, by `parseFloat` for example, would take a while, if accepted at all (see [Check if Napi::Value is integer number?](https://github.com/nodejs/node-addon-api/issues/265) or [More special types check](https://github.com/nodejs/node-addon-api/issues/57) ).

Python RFC wrapper does not have this limitation and [returns Python Decimal object](https://github.com/SAP/PyRFC/blob/master/src/pyrfc/_pyrfc.pyx#L2076) because Cython makes practically all Python functionality available here.

What is the biggest practical problem with current node-rfc behaviour, accepting string, number or Decimal object as BCD input and returning only string as output? Parsing on application side is not slower and not always required. For web applications CRUD cycles, the number or string format does not not matter, at least in my web projects with node-rfc. 

What is the biggest pain point in your scenario/application ?

**Comment by mploigtweber** [comment]
my problem is: at each point where a BAPI or an RFC function outputs a numeric value, text is now output by the node-rfc module. For example, we have several programs here where the JSON is parsed by the node.js webservice and the classes expect numeric values.

If we wanted to use the node-rfc module here, we would have to change code in many places. It just does not feel natural. The SAP documentation says, here comes a number and text comes out on the other side. Maybe we are just spoiled by the node-sapnwrfc module here.

If you now say that the output as a string is now much more accurate or just free of errors, then I have to bite into the bullet and overwork the affected areas with parseFloat.

I thought it would be relatively easy to make the output optional.

And as far as I've read, the issue of BCDs was numerical recently.

Thank you

**Comment by bsrdjan** [comment]
yes, the BCD was numerical but inaccurate, but can switch back to that if you like :) ?

Here the old [set/fill](https://github.com/SAP/node-rfc/blob/0.10-api/src/rfcio.cc#L161) and [get/wrap](https://github.com/SAP/node-rfc/blob/0.10-api/src/rfcio.cc#L389) logic, using Floats for both ABAP BCD and ABAP FLOAT types. The BCD "fall-through" to FLOAT was a bug, fixed after rounding  issues reported with BCD.

Let keep it open and I will analyse if solution can be provided, ideally wihout napi change requests. The fix, if possible, would bringing the node-rfc BCD handling "in sync" with PyRFC and JCo.

This week I have not much capacity and next 2 weeks absent. In which timeframe would you need a definitive answer/solution?

**Comment by mploigtweber** [comment]
Do not stress. We have a running system here with node.js (jdorner ... node version 5.0) and dotNet connector. But I would like to stay exclusively at node.js and switch to the node-rfc module when version 1.0.0 is released. Then we could finally update to an LTS version of node.js. If there were an output option in the future, that would be great.

Thanks in advance :-)

**Comment by bsrdjan** [comment]
The simple solution should work in just released @next rc5. BCD and FLOAT both accept string, number or Decimal object and return number. Darwin and node 11 are not stable yet.

**Comment by mploigtweber** [comment]
Very cool. Just tried. It will be outputed again numbers :-)

What do you think? What should the final version look like? Will there be an optional parameter?

Thank you very much and Happy Halloween

**Comment by soulchild** [comment]
> BCD and FLOAT both accept string, number or Decimal object and return number. 

Does this mean, that the return type of BCD values is now again a float instead of a string or am I misunderstanding you?

**Comment by bsrdjan** [comment]
> Does this mean, that the return type of BCD values is now again a float instead of a string or am I misunderstanding you?

Regardless if BCD values returned as strings or floats, from ABAP to nodejs, any follow-up calculations in nodejs would require a conversion to Decimal.js or similar object, to prevent rounding issues:

```shell
> x =0.1 # received as float from ABAP, or as a string '0.1', and in nodejs parsed as float
0.1
> x.toPrecision(55)
'0.1000000000000000055511151231257827021181583404541015625'
> y = 0
0
> for (let i = 0; i < 10000000; i++) y+= x;
999999.9998389754
```

with Decimal

```shell
> D = require('Decimal.js')
> x = D(.1)
> y = D(0)
> for (let i = 0; i < 10000000; i++) y = y.add(x);
> y.toNumber().toPrecision(55)
'1000000.000000000000000000000000000000000000000000000000'
>
```

In this context, returning ABAP BCD as node-rfc string or float, I see both equally good for the representation and both equally wrong for calculations and more the question of convention and convenience for one or another application? 

Do you see it the same way?

The optimal solution would be perhaps a node-rfc configuration option, to provide a pointer to JS Decimal function, for ABAP BCD values conversion. Without configuration, which is standard only because no standard decimal exists in JavaScript, use the string (not recommended standard), providing options to use the float (not recommended option), or decimal object (recommended option). Would that work for all?

**Comment by mploigtweber** [comment]
Sounds like a good approach. From my side, a configurable solution with decimal.js behavior would be optimal.

Thanks in advance

**Comment by soulchild** [comment]
Sounds good to me as well 👍

**Comment by bsrdjan** [comment]
Please check the prerelease 6, just published and these [unit tests](https://github.com/SAP/node-rfc/blob/napi/test/datatypes.spec.js#L43)

**Comment by mploigtweber** [comment]
Hi,

again on the topic. I have now installed the version node-rfc@1.0.0-rc10 on a new server. Now I was hoping to use the BCD option (still need 'number' as output). From what I have found so I get only the variant without pool with the option bcd: "number" started.

`const rfcClient = require ('node-rfc'). Client;

let client = new rfcClient (abapSystemP200, {
**bcd: "number"**
});

await client.open ();
let result = await client.call (bapi, paras);
await client.close ();
return result; `


I would need this functionality but also with the connection pool. Did I miss something?

Thanks in advance

**Comment by bsrdjan** [comment]
Thanks @mploigtweber for catching this bug. I opened the new issue for tracking: #104

**Comment by bsrdjan** [comment]
Fixed by ac6eb8883b95d7f9aae3c22440578d95edc6ab51, will be provided in next release.

**Comment by mploigtweber** [comment]
Hello @bsrdjan ... as I see it, the client options for BCD representation in the pool class have not arrived in version 1.01 again ☹

**Comment by bsrdjan** [comment]
should be back again in just published 1.0.2

**Comment by mploigtweber** [comment]
> should be back again in just published 1.0.2

good work :-)
